### PR TITLE
[MRG] POT numpy/torch/jax backends 

### PR DESCRIPTION
**Pull request by rflamary**
## Types of changes

We are working on a new multi-backend API for POT with the objective to be able to handle numpy/torch/jax/cupy arrays seamlessly.

This will be a big one.

Expected features :

- [x] Numpy backend
- [x] Torch backend + tests on CPU
- [x] Jax backend + tests on CPU
- [x] ot.dist working with numpy/jax/torch arrays
- [x] ot.emd working with numpy/jax/torch arrays
- [x] ot.emd2 working with numpy/jax/torch arrays (with working gradients wrt a,b,M except jax that require to define a new function)
- [x] ot.sinkhorn working with numpy/jax/torch arrays (default algorithm only)
- [x] ot.sinkhorn2 working with numpy/jax/torch arrays (with working gradients wrt a,b,M)
- [x] Update documentations from emd/emd2 sinkhorn/sinkhorn2 and quickstart guide for the new backend
- [x] ot.utils.proj_simplex working with numpy/jax/torch arrays

### CODE REVIEW

**Comment by rflamary** [comment]
Just updated the doc one last time, it should build ;).

The PR feels OK, wdyt @ncourty ?

File: docs/source/quickstart.rst [hunk 1]
```diff
@@ -950,6 +962,60 @@ explicitly.
     use it you have to specifically import it with :code:`import ot.gpu` .
 
 
+Solving OT with Multiple backends
+---------------------------------
+
+.. _backends_section:
+
+Since version 0.8, POT provides a backend that allows to code solvers
+independent from the type of the input arrays. The idea is to provide the user
```
Comment (agramfort) [inline docs/source/quickstart.rst:971]:
```suggestion
independently from the type of the input arrays. The idea is to provide the user
```

File: docs/source/quickstart.rst [hunk 2]
```diff
@@ -950,6 +962,60 @@ explicitly.
...
+with a toolbox that works seamlessly and returns a solution for instance as a
```
Comment (agramfort) [inline docs/source/quickstart.rst:972]:
```suggestion
with a package that works seamlessly and returns a solution for instance as a
```
toolbox sounds very matlabish :)

File: docs/source/quickstart.rst [hunk 3]
```diff
@@ -950,6 +962,60 @@ explicitly.
...
+
+.. code:: python
+
+    # a and b are 1D histograms (sum to 1 and positive)
+    # M is the ground cost matrix
+    T = ot.emd(a, b, M)  # exact linear program
+    w = ot.emd2(a, b, M)  # Wasserstein computation
+
+the functions  :any:`ot.emd` and :any:`ot.emd2` can take inputs of the type
+:any:`numpy.array` :any:`torch.tensor` or  :any:`jax.numpy.array`. The output of
```
Comment (agramfort) [inline docs/source/quickstart.rst:992]:
```suggestion
:any:`numpy.array`, :any:`torch.tensor` or  :any:`jax.numpy.array`. The output of
```

File: docs/source/readme.rst [hunk 4]
```diff
@@ -206,36 +214,36 @@ Short examples
...
 
-   .. code:: python
+.. code:: python
 
-       # a,b are 1D histograms (sum to 1 and positive)
-       # M is the ground cost matrix
-       Wd=ot.emd2(a,b,M) # exact linear program
-       Wd_reg=ot.sinkhorn2(a,b,M,reg) # entropic regularized OT
-       # if b is a matrix compute all distances to a and return a vector
+    # a,b are 1D histograms (sum to 1 and positive)
```
Comment (agramfort) [inline docs/source/readme.rst:225]:
```suggestion
    # a and b are 1D histograms (sum to 1 and positive)
```

File: docs/source/readme.rst [hunk 5]
```diff
@@ -206,36 +214,36 @@ Short examples
...
 -  Compute OT matrix
 
-   .. code:: python
+.. code:: python
 
-       # a,b are 1D histograms (sum to 1 and positive)
-       # M is the ground cost matrix
-       T=ot.emd(a,b,M) # exact linear program
-       T_reg=ot.sinkhorn(a,b,M,reg) # entropic regularized OT
+    # a,b are 1D histograms (sum to 1 and positive)
```
Comment (agramfort) [inline docs/source/readme.rst:235]:
```suggestion
    # a and b are 1D histograms (sum to 1 and positive)
```

File: ot/backend.py [hunk 6]
```diff
@@ -0,0 +1,534 @@
...
+        Also raises TypeError if all arrays are not from the same backend
+    """
+    # check that some arrays given
+    if not len(args) > 0:
+        raise ValueError(" The function takes at least one parameter")
+    # check all same type
+
+    if isinstance(args[0], np.ndarray):
+        if not len(set(type(a) for a in args)) == 1:
+            raise ValueError("All array should be from the same type/backend. Current types are : {}".format([type(a) for a in args]))
```
Comment (agramfort) [inline ot/backend.py:54]: I am surprised the CIs don't complain about these long lines.

File: ot/bregman.py [hunk 7]
```diff
@@ -339,14 +348,19 @@ def sinkhorn_knopp(a, b, M, reg, numItermax=1000,
...
 
-    a = np.asarray(a, dtype=np.float64)
-    b = np.asarray(b, dtype=np.float64)
-    M = np.asarray(M, dtype=np.float64)
+    if isinstance(a, list):
+        a = np.array(a)
+    if isinstance(b, list):
+        b = np.array(b)
+    if isinstance(M, list):
+        M = np.array(M)
```
Comment (agramfort) [inline ot/bregman.py:356]: I would suggest a check_array function like in sklearn

```
a, b, M = check_array(a, b, M)
```

File: ot/bregman.py [hunk 8]
```diff
@@ -363,21 +377,22 @@ def sinkhorn_knopp(a, b, M, reg, numItermax=1000,
...
     # print(reg)
 
     # Next 3 lines equivalent to K= np.exp(-M/reg), but faster to compute
-    K = np.empty(M.shape, dtype=M.dtype)
-    np.divide(M, -reg, out=K)
-    np.exp(K, out=K)
+    #K = np.empty(M.shape, dtype=M.dtype)
+    #np.divide(M, -reg, out=K)
+    #np.exp(K, out=K)
+    K = nx.exp(M / (-reg))
```
Comment (agramfort) [inline ot/bregman.py:392]: cleanup? cf comment above

File: ot/lp/__init__.py [hunk 9]
```diff
@@ -189,37 +189,41 @@ def emd(a, b, M, numItermax=100000, log=False, center_dual=True):
...
         :ref:`center_ot_dual`.
 
     Returns
     -------
-    gamma: (ns x nt) numpy.ndarray
-        Optimal transportation matrix for the given parameters
-    log: dict
-        If input log is true, a dictionary containing the cost and dual
-        variables and exit status
+    gamma: (ns x nt) array-like
```
Comment (agramfort) [inline ot/lp/__init__.py:221]:
```suggestion
    gamma: array-like, shape (ns, nt) 
```
is the recommended way. It should also be consistent everywhere :)

File: ot/lp/__init__.py [hunk 10]
```diff
@@ -232,26 +236,42 @@ def emd(a, b, M, numItermax=100000, log=False, center_dual=True):
...
+    ot.bregman.sinkhorn : Entropic regularized OT ot.optim.cg : General
+    regularized OT"""
+
+    # convert to numpy if list
+    if type(a)==list:
+        a=np.array(a)
+    if type(b)==list:
+        b=np.array(b)   
+    if type(M)==list:
+        M=np.array(M)
```
Comment (agramfort) [inline ot/lp/__init__.py:261]: check_array...

also you should use isinstance and not type(...) ==

File: ot/lp/__init__.py [hunk 11]
```diff
@@ -262,6 +282,10 @@ def emd(a, b, M, numItermax=100000, log=False, center_dual=True):
     assert (a.shape[0] == M.shape[0] and b.shape[0] == M.shape[1]), \
         "Dimension mismatch, check dimensions of M with a and b"
 
+    # ensure that same mass
+    np.testing.assert_almost_equal(a.sum(0),b.sum(0),err_msg='a and b vector must have the same sum')
```
Comment (agramfort) [inline ot/lp/__init__.py:286]: pep8

I thought we had pep8 enforced in the project

File: ot/lp/__init__.py [hunk 12]
```diff
@@ -367,12 +390,27 @@ def emd2(a, b, M, processes=multiprocessing.cpu_count(),
     ot.bregman.sinkhorn : Entropic regularized OT
     ot.optim.cg : General regularized OT"""
 
+    if type(a)==list:
+        a=np.array(a)
+    if type(b)==list:
+        b=np.array(b)   
+    if type(M)==list:
+        M=np.array(M)
```
Comment (agramfort) [inline ot/lp/__init__.py:398]: check_array...

File: ot/lp/__init__.py [hunk 13]
```diff
@@ -418,6 +458,9 @@ def f(b):
             if np.any(~asel) or np.any(~bsel):
                 u, v = estimate_dual_null_weights(u, v, a, b, M)
 
+            G = nx.from_numpy(G, type_as=M0)
+            cost = nx.set_gradients(nx.from_numpy(cost, type_as=M0),(a0,b0, M0),(nx.from_numpy(u, type_as=a0),nx.from_numpy(v, type_as=b0),G))
```
Comment (agramfort) [inline ot/lp/__init__.py:462]: lines less than 80 or at most 100 please :)

File: ot/utils.py [hunk 14]
```diff
@@ -84,60 +139,82 @@ def euclidean_distances(X, Y, squared=False):
...
+        'euclidean', 'hamming', 'jaccard', 'kulsinski', 'mahalanobis',
+        'matching', 'minkowski', 'rogerstanimoto', 'russellrao', 'seuclidean',
         'sokalmichener', 'sokalsneath', 'sqeuclidean', 'wminkowski', 'yule'.
 
 
     Returns
     -------
 
-    M : np.array (n1,n2)
+    M : array-like (n1,n2)
```
Comment (agramfort) [inline ot/utils.py:203]:
```suggestion
    M : array-like, shape (n1, n2)
```

File: test/test_backend.py [hunk 15]
```diff
@@ -0,0 +1,351 @@
...
+
+def test_get_backend_list():
+
+    lst = get_backend_list()
+
+    assert len(lst) > 0
+    assert isinstance(lst[0], ot.backend.NumpyBackend)
+
+
+def test_to_numpy():
```
Comment (agramfort) [inline test/test_backend.py:30]: use pytest parametrize

File: test/test_backend.py [hunk 16]
```diff
@@ -0,0 +1,351 @@
...
+            get_backend(A, B2)
+
+
+@pytest.mark.parametrize('nx', backend_list)
+def test_convert_between_backends(nx):
+
+    A = np.zeros((3, 2))
+    B = np.zeros((3, 1))
+
+    for nx in get_backend_list()[1:]:
```
Comment (agramfort) [inline test/test_backend.py:85]: you overwrite here the parametrize nx argument

File: test/test_backend.py [hunk 17]
```diff
@@ -0,0 +1,351 @@
...
+def test_func_backends(backend):
+
+    rnd = np.random.RandomState(0)
+    M = rnd.randn(10, 3)
+    v = rnd.randn(3)
+    val = np.array([1.0])
+
+    lst_tot = []
+
+    for nx in [ot.backend.NumpyBackend(), backend]:
```
Comment (agramfort) [inline test/test_backend.py:177]: NumpyBackend  is not already in backend_list? if not so why?

File: test/test_backend.py [hunk 18]
```diff
@@ -0,0 +1,351 @@
...
+        assert np.allclose(a1, a2, atol=1e-7)
+
+
+def test_gradients_backends():
+
+    rnd = np.random.RandomState(0)
+    v = rnd.randn(10)
+    c = rnd.randn(1)
+
+    if torch:
```
Comment (agramfort) [inline test/test_backend.py:337]: the fact that you have to branch for torch here suggests that the backend mechanism
is not consistent

File: test/test_utils.py [hunk 19]
```diff
@@ -158,14 +203,14 @@ def test_check_params():
...
-    @ot.utils.deprecated('deprecated text for fun')
+    @ ot.utils.deprecated('deprecated text for fun')
     def fun():
         pass
 
     def fun2():
         pass
 
-    @ot.utils.deprecated('deprecated text for class')
+    @ ot.utils.deprecated('deprecated text for class')
```
Comment (agramfort) [inline test/test_utils.py:213]:
```suggestion
    @ot.utils.deprecated('deprecated text for class')
```

File: test/test_utils.py [hunk 20]
```diff
@@ -158,14 +203,14 @@ def test_check_params():
 
 def test_deprecated_func():
 
-    @ot.utils.deprecated('deprecated text for fun')
+    @ ot.utils.deprecated('deprecated text for fun')
```
Comment (agramfort) [inline test/test_utils.py:206]:
```suggestion
    @ot.utils.deprecated('deprecated text for fun')
```

File: test/test_backend.py (hunk 17 above)
Comment (rflamary) [reply test/test_backend.py:177]: because we run it for numpy and the current backend and check that they provide the same output (see below the loop)

File: test/test_backend.py (hunk 18 above)
Comment (rflamary) [reply test/test_backend.py:337]: for gradients no we cannot be consistent because each backend handes gradients differently. The fact that we have a branch mans that we can handle different ways to compute the gradients (when they can be computed).

if you want to test if the backend works, you have to do a test utiings functiosn specfis to ecah backend.

File: ot/backend.py (hunk 6 above)
Comment (rflamary) [reply ot/backend.py]: me too, flake8 says nothing...

Comment (agramfort) [reply ot/backend.py]: maybe due to 
https://github.com/PythonOT/POT/blob/master/setup.cfg#L6

File: examples/backends/plot_unmix_optim_torch.py [hunk 21]
```diff
@@ -0,0 +1,158 @@
+# -*- coding: utf-8 -*-
+"""
+=================================
+Wasserstein unmixing with Pytorch
+=================================
+
+In this example we estimate mixing parameters from distributions that minimize
+the Wassersytein distance. In other words we suppose that a target
```
Comment (agramfort) [inline examples/backends/plot_unmix_optim_torch.py]:
```suggestion
the Wasserstein distance. In other words we suppose that a target 
```

File: examples/backends/plot_unmix_optim_torch.py [hunk 22]
```diff
@@ -0,0 +1,158 @@
...
+distribution simplex :math:`\Delta_K`.
+
+In order to estimate this weight vector we propose to optimize the Wasserstein 
+distance between the model and the observed :math:`\mu^t` with respect to 
+the vector. This leads to the following optimization problem:
+    
+.. math:: 
+    \min_{\mathbf{w}\in\Delta_K`} \quad W(\mu^t,\sum_{k=1}^K w_k\mu^s_k)
+
+This minimization is done in this example with a  simple projected gradient
```
Comment (agramfort) [inline examples/backends/plot_unmix_optim_torch.py]:
```suggestion
This minimization is done in this example with a simple projected gradient 
```

File: examples/backends/plot_unmix_optim_torch.py [hunk 23]
```diff
@@ -0,0 +1,158 @@
...
+
+In order to estimate this weight vector we propose to optimize the Wasserstein 
+distance between the model and the observed :math:`\mu^t` with respect to 
+the vector. This leads to the following optimization problem:
+    
+.. math:: 
+    \min_{\mathbf{w}\in\Delta_K`} \quad W(\mu^t,\sum_{k=1}^K w_k\mu^s_k)
+
+This minimization is done in this example with a  simple projected gradient 
+descent in Pytorch. We use the automatic backend of POT that allows us to
```
Comment (agramfort) [inline examples/backends/plot_unmix_optim_torch.py]:
```suggestion
descent in PyTorch. We use the automatic backend of POT that allows us to 
```

File: examples/backends/plot_unmix_optim_torch.py [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+In order to estimate this weight vector we propose to optimize the Wasserstein 
+distance between the model and the observed :math:`\mu^t` with respect to 
+the vector. This leads to the following optimization problem:
+    
+.. math:: 
+    \min_{\mathbf{w}\in\Delta_K`} \quad W(\mu^t,\sum_{k=1}^K w_k\mu^s_k)
+
+This minimization is done in this example with a  simple projected gradient 
+descent in Pytorch. We use the automatic backend of POT that allows us to 
+compute teh Wasserstein distance with :any:`ot.emd2` with
```
Comment (agramfort) [inline examples/backends/plot_unmix_optim_torch.py:27]:
```suggestion
compute the Wasserstein distance with :any:`ot.emd2` with
```

File: examples/backends/plot_unmix_optim_torch.py [hunk 25]
```diff
@@ -0,0 +1,158 @@
...
+pl.title('Sources and Target distributions')
+pl.legend()
+
+
+##############################################################################
+# Optimization of the model wrt the Wasserstein distance
+# ------------------------------------------------------
+
+
+#%% barycenter optimiation with gradient decent
```
Comment (agramfort) [inline examples/backends/plot_unmix_optim_torch.py:98]:
```suggestion
#%% barycenter optimization with gradient descent
```

File: examples/backends/plot_unmix_optim_torch.py [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+w = torch.tensor(ot.unif(2), requires_grad=True)
+
+# uniform weights for target
+b = torch.tensor(ot.unif(nt))
+
+lr = 2e-3  # learning rate
+niter = 500  # number of iterations
+losses = []  # loss along the iterations
+
+# loss for the minimal wasrsetsien estimator
```
Comment (agramfort) [inline examples/backends/plot_unmix_optim_torch.py:114]:
```suggestion
# loss for the minimal Wasserstein estimator
```

File: examples/backends/plot_unmix_optim_torch.py [hunk 27]
```diff
@@ -0,0 +1,158 @@
...
+def get_loss(w):
+    a = torch.mv(H2, w)  # distribution reweighting
+    return ot.emd2(a, b, M2) # squared Wassersein 2
```
Comment (agramfort) [inline examples/backends/plot_unmix_optim_torch.py]:
```suggestion
    return ot.emd2(a, b, M2)  # squared Wasserstein 2 
```

**Comment by rflamary** [comment]
here is the new example. I love how simply the Wasserstein loss can be used and optimized.

https://626-71472695-gh.circle-artifacts.com/0/dev/auto_examples/backends/plot_unmix_optim_torch.html#sphx-glr-auto-examples-backends-plot-unmix-optim-torch-py

Just corrected the legend in the last figure. Will merge when tests pass.

**Comment by agramfort** [comment]
really cool !

>
